src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                lock (this) {
+                    foreach (var disposeScope in DisposeScopeStack) {
+                        size += disposeScope.TotalAllocatedSize;
+                    }
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            Interlocked.Increment(ref _createdCount);
+
+            lock (this) {
+                if (DisposeScopeStack.Count > 0) {
+                    DisposeScopeStack.Peek().Include(disposable);
+                }
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            Interlocked.Increment(ref _disposedCount);
+            lock (this) {
+                foreach (var disposeScope in DisposeScopeStack) {
+                    disposeScope.Disposables.Remove(disposable);
+                }
+            }
+        }
+
+        internal static DisposeScope NewDisposeScope()
+        {
+            lock (Singleton) {
+                return Singleton.InnerNewDisposeScope();
+            }
+        }
+
+        private DisposeScope InnerNewDisposeScope()
+        {
+            lock (this) {
+                var disposeScope = new DisposeScope(this);
+                DisposeScopeStack.Push(disposeScope);
+                return disposeScope;
+            }
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                Debug.Assert(DisposeScopeStack.Count > 0);
+                Debug.Assert(DisposeScopeStack.Peek() == disposeScope);
+                DisposeScopeStack.Pop();
+            }
+        }
+
+        internal void MoveToOuterScope(DisposeScope disposeScope, IDisposable disposable)
+        {
+            lock (this) {
+                var array = DisposeScopeStack.ToArray();
+                for (var i = array.Length - 1; i >= 1; i--) {
+                    if (array[i] == disposeScope) {
+                        array[i - 1].Include(disposable);
+                        return;
+                    }
+                }
+            }
+        }
+
+        public class DisposeScope : IDisposable
+        {
+            private readonly DisposeScopeManager _disposeScopeManager;
+
+            public DisposeScope(DisposeScopeManager disposeScopeManager)
+            {
+                _disposeScopeManager = disposeScopeManager;
+            }
+
+            /// <summary>
+            /// The disposables that are scheduled for disposing.
+            /// </summary>
+            public HashSet<IDisposable> Disposables { get; } = new();
```

Comment:
I think this set should be using [`ReferenceEqualityComparer`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.referenceequalitycomparer?view=net-6.0)

Reply:
That doesn't seem to exist in the version of C# I'm using, it seems like it's a .NET 6. I'll try to use something similar until we've migrated to .NET 6?

Reply:
@mfagerlund it should be trivial to reimplement using [`RuntimeHelpers.GetHashCode`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.gethashcode) and `object.ReferenceEquals`

See (actual implementation in .NET 5.0+)[https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/System/Collections/Generic/ReferenceEqualityComparer.cs]

Reply:
Yep, I found a similar one that I'm using now.